src/unused_finder/export_collector.rs

```diff
@@ -0,0 +1,477 @@
+use std::{
+    collections::{HashMap, HashSet},
+    iter::FromIterator,
+    sync::Arc,
+};
+
+use swc_core::{
+    common::{
+        comments::{CommentKind, Comments, SingleThreadedComments},
+        BytePos, Spanned,
+    },
+    ecma::{
+        ast::{
+            BindingIdent, CallExpr, Callee, Decl, ExportAll, ExportDecl, ExportDefaultDecl,
+            ExportDefaultExpr, ExportSpecifier, Id, ImportDecl, ImportSpecifier, Lit,
+            ModuleExportName, NamedExport, Pat, Str, TsImportEqualsDecl,
+        },
+        visit::{Visit, VisitWith},
+    },
+};
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub enum ExportedItem {
+    Named(String),
+    Default,
+    Namespace,
+    ExecutionOnly, // in case of `import './foo';` this executes code in file but imports nothing
+}
+
+impl From<&ImportedItem> for ExportedItem {
+    fn from(i: &ImportedItem) -> Self {
+        match i {
+            ImportedItem::Named(named) => ExportedItem::Named(named.clone()),
+            ImportedItem::Default => ExportedItem::Default,
+            ImportedItem::Namespace => ExportedItem::Namespace,
+            ImportedItem::ExecutionOnly => ExportedItem::ExecutionOnly,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub enum ImportedItem {
+    Named(String),
+    Default,
+    Namespace,
+    ExecutionOnly, // in case of `import './foo';` this executes code in file but imports nothing
+}
+
+impl From<&ExportedItem> for ImportedItem {
+    fn from(e: &ExportedItem) -> Self {
+        match e {
+            ExportedItem::Named(name) => ImportedItem::Named(name.clone()),
+            ExportedItem::Default => ImportedItem::Default,
+            ExportedItem::Namespace => ImportedItem::Namespace,
+            ExportedItem::ExecutionOnly => ImportedItem::ExecutionOnly,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ExportsCollector {
+    // `import foo, {bar as something} from './foo'` generates `{ "./foo": ["default", "bar"] }`
+    pub imported_ids_path_name: HashMap<String, HashSet<ImportedItem>>,
+    // require('foo') generates ['foo']
+    pub require_paths: HashSet<String>,
+    // import('./foo') and import './foo' generates ["./foo"]
+    pub imported_paths: HashSet<String>,
+    // `export {default as foo, bar} from './foo'` generates { "./foo": ["default", "bar"] }
+    pub export_from_ids: HashMap<String, HashSet<ImportedItem>>,
+    pub exported_ids: HashSet<ExportedItem>,
+    // import './foo';
+    pub executed_paths: HashSet<String>,
+    // exported from this file
+    // const foo = require('foo') generates ["foo"]
+    require_identifiers: HashSet<Id>,
+    skipped_items: Arc<Vec<regex::Regex>>,
+    pub comments: SingleThreadedComments,
```

Comment:
I think this means we end up storing a bunch of AST information in the collector, which seems like a bad idea, since it means we retain a lot of memory on the heap.

Instead, which exports are allowed to be ignored should probably be calculated during the traversal, and then discarded.

This could be stored as a hashmap within the Collector (ignoredExports)?

Reply:
```rust
struct ExportsData {
   // our existing data fields
   
    pub export_from_ids: HashMap<String, HashSet<ImportedItem>>,
    pub exported_ids: HashSet<ExportedItem>,
    
    // etc...

}

struct ExportsCollector  {
  exportsData: ExportsData,
  skipped_items: ...,
  comments: SingleThreadedComments,
}

///discard ExportsCollector after traversal, return ExportsData
```

Reply:
This functionality already exists in `unused_finder_visitor_runner.rs`